pubsub/v1/handle.go

```diff
@@ -0,0 +1,129 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pubsub
+
+import (
+	"context"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/dapr-sandbox/components-go-sdk/internal"
+	contribPubSub "github.com/dapr/components-contrib/pubsub"
+	proto "github.com/dapr/dapr/pkg/proto/components/v1"
+	"github.com/pkg/errors"
+)
+
+var ErrAckTimeout = errors.New("ack has timed out")
+
+// threadSafeStream wraps a grpc stream with locks to permit send and recv in multiples goroutines.
+type threadSafeStream interface {
+	send(*proto.PullMessagesResponse) error
+	recv() (*proto.PullMessagesRequest, error)
+}
+
+// grpcThreadSafeStream wraps a grpcStream for thread safe operations.
+// As per documentation is unsafe to call recv OR send in multiples goroutines
+// https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md#streams
+// it is safe to call recv AND send in different goroutines.
+type grpcThreadSafeStream struct {
+	recvLock *sync.Mutex
+	sendLock *sync.Mutex
+	stream   proto.PubSub_PullMessagesServer
+}
+
+func (s *grpcThreadSafeStream) send(msg *proto.PullMessagesResponse) error {
+	s.sendLock.Lock()
+	defer s.sendLock.Unlock()
+	return s.stream.Send(msg)
+}
+
+func (s *grpcThreadSafeStream) recv() (*proto.PullMessagesRequest, error) {
+	s.recvLock.Lock()
+	defer s.recvLock.Unlock()
+	return s.stream.Recv()
+}
+
+// ackLoop starts a active ack loop reciving acks from client.
+func ackLoop(tfStream threadSafeStream, ackManager *acknowledgementManager) error {
+	for {
+		ack, err := tfStream.recv()
+		if err == io.EOF {
+			return nil
+		}
+
+		if err != nil {
+			// FIXME
+			// should we continue without sleep ?
+			// should we stop and cancel everything?
+			pubsubLogger.Errorf("error %v when trying to receive ack, sleeping 5 seconds", err)
+			time.Sleep(time.Second * 5)
+			continue
+		}
+
+		var ackError error
+
+		if ack.AckError != nil {
+			ackError = errors.New(ack.AckError.Message)
+		}
+		if err := ackManager.ack(ack.AckMessageId, ackError); err != nil {
+			pubsubLogger.Warnf("error %v when trying to notify ack", err)
+		}
+	}
+}
+
+// handler build a pubsub handler using the given threadsafe stream and the ack manager.
+func handler(tfStream threadSafeStream, ackManager *acknowledgementManager) contribPubSub.Handler {
+	return func(ctx context.Context, contribMsg *contribPubSub.NewMessage) error {
+		msgID, pendingAck, cleanup := ackManager.get()
+		defer cleanup()
+
+		msg := &proto.PullMessagesResponse{
+			Data:        contribMsg.Data,
+			TopicName:   contribMsg.Topic,
+			Metadata:    contribMsg.Metadata,
+			ContentType: internal.ZeroIfNil(contribMsg.ContentType),
```

Comment:
Zero? For contenttype?

Reply:
Zero is how go call variables declared without an explicit initial value https://go.dev/tour/basics/12

Reply:
If "zero" as default value is a Go-ism, cool. But when I read this line, my expectation is that a literal zero will be returned here.

Reply:
How about ZeroValueIfNil?